### ISSUE TITLE
Conditional timed event trigger

### ISSUE THREAD

**Issue by asad-awadia** [Gurpartap/statemachine-go#2]
Is it possible to add a conditional timed fire? Like if current state is x or not x then trigger it 
This will avoid piling up triggered goroutines if no need to trigger

@Gurpartap  - this project is very involved thanks for the comprehensive library

**Comment by Gurpartap** [comment]
it's already possible with `TimedEvery(duration time.Duration)` + `Choice(...) .OnTrue() .OnFalse()`

the 'tick' event usage in the cognizant example exhibits the use case rather well. add a transition/event callback, if necessary.

**Comment by asad-awadia** [comment]
@Gurpartap yes I saw that example but a bit confused

1. is the choice function there to determine whether the tick will happen or not? i.e. has the tick already been scheduled and it is just a boolean for the ontrue on false functins?
2. What state is the state machine in when we are in the choice function?

I want to only trigger the tick only if the current state is 'idle'  - i am not sure if choice will allow me to do that

**Comment by Gurpartap** [comment]
> 1. is the choice function there to determine whether the tick will happen or not? i.e. has the tick already been scheduled and it is just a boolean for the ontrue on false functins?

the tick event triggers a transition defined by `Choice`, which is just a glorified `if {} else {}`.

in other words, on `TimedEvery(...)`, tick event will always happen (unless the state machine is busy processing another event/transition/callback – or `TimedEvery` is being paused manually using `SkipUntil(...)`). 

once tick event is fired, it'll call `Choice(...)`, which can define transitions based on whether the result of Choice was true or not.

> 2. What state is the state machine in when we are in the choice function?

events don't change the state. events are just for triggering a transition. 

transitions cause the state to change. if there is no transition, nothing happens.

in other words, the state will only change when a choice result triggers a transition successfully.

> I want to only trigger the tick only if the current state is 'idle'  - i am not sure if choice will allow me to do that

if you only want the machine to change state automatically when it becomes idle, one of the way to do it with TimedEvery and without Choice would be as simple as:

```go
m.Event("tick").
  TimedEvery(1 * time.Second).
  Transition().From("idle").To("processing")
```

another way is to handle the state becoming idle outside the state machine definition by using a callback. this would be more "immediate" than the TimedEvery tick, which can get delayed by about a second.

```go
m.AfterTransition().FromAny().To("idle").Do(someFunc)
```

**Comment by asad-awadia** [comment]
```
m.Event("tick").
  TimedEvery(1 * time.Second).
  Transition().From("idle").To("processing")
```

in this example it only creates a tick event if the current state is idle? and nothing happens if the current state is something else?

What happens in terms of threading/concurrency if another non-tick event is sent at the same time as that tick gets fired?

**Comment by Gurpartap** [comment]
> ```
> 
> m.Event("tick").
> 
>   TimedEvery(1 * time.Second).
> 
>   Transition().From("idle").To("processing")
> 
> ```
> 
> 
> 
> in this example it only creates a tick event if the current state is idle? and nothing happens if the current state is something else?

TimedEvery is like a cron job. it fires the tick event every second in that example. on every tick event, if current state is "idle" it will transition to "processing". you can add a transition callback to trigger some func in your code when the transition occurs.

> What happens in terms of threading/concurrency if another non-tick event is sent at the same time as that tick gets fired?

`machine.Fire("event")` is blocking. each event is handled one after the other. the current event (and the resulting transition, if any) would have to finish before the next event trigger is handled.

**Comment by asad-awadia** [comment]
hmm ok ok

Will have to try out various things

thanks
